internal/graph/singleflight_resolver_test.go

```diff
@@ -0,0 +1,174 @@
+package graph
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/oklog/ulid/v2"
+	openfgav1 "github.com/openfga/api/proto/openfga/v1"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/openfga/openfga/pkg/storage/memory"
+	"github.com/openfga/openfga/pkg/storage/storagewrappers"
+	"github.com/openfga/openfga/pkg/testutils"
+	"github.com/openfga/openfga/pkg/tuple"
+	"github.com/openfga/openfga/pkg/typesystem"
+)
+
+func TestSingleflightResolver(t *testing.T) {
+	ds := memory.New()
+	defer ds.Close()
+
+	storeID := ulid.Make().String()
+
+	var tuples = []*openfgav1.TupleKey{
+		tuple.NewTupleKey("folder:1", "viewer", "user:jon"),
+
+		tuple.NewTupleKey("folder:2", "parent1", "folder:1"),
+		tuple.NewTupleKey("folder:2", "parent2", "folder:1"),
+
+		tuple.NewTupleKey("folder:3", "parent1", "folder:2"),
+		tuple.NewTupleKey("folder:3", "parent2", "folder:2"),
+	}
+
+	err := ds.Write(context.Background(), storeID, nil, tuples)
+	require.NoError(t, err)
+
+	model := testutils.MustTransformDSLToProtoWithID(`model
+	schema 1.1
+
+	type user
+
+	type folder
+	relations
+	  define parent1: [folder]
+	  define parent2: [folder]
+	  define viewer: [user] or viewer from parent1 or viewer from parent2 or viewer`)
+
+	ctx := typesystem.ContextWithTypesystem(context.Background(), typesystem.New(model))
+
+	checker := NewLocalChecker(
+		storagewrappers.NewCombinedTupleReader(ds, []*openfgav1.TupleKey{}),
+	)
+	defer checker.Close()
+
+	resp, err := checker.ResolveCheck(ctx, &ResolveCheckRequest{
+		StoreID:            storeID,
+		TupleKey:           tuple.NewTupleKey("folder:3", "viewer", "user:jon"),
+		ContextualTuples:   nil,
+		ResolutionMetadata: &ResolutionMetadata{Depth: defaultResolveNodeLimit},
+	})
+
+	require.NoError(t, err)
+	require.True(t, resp.GetAllowed())
```

Comment:
Shouldn't these be asserts? Except for the above require?

Reply:
Not a current practice in this codebase.

Reply:
It's not really about practice, it's about the fact that you have a series of assertions and you want to immediately see what passes and what fails, instead of stopping at the first one that fails.

Reply:
Our assertions come from the `require` package. That's the current practice. Maybe I misunderstand what you're saying @sergiught . Could you be explicit with an example demonstrating what you mean?

Reply:
@jon-whit , `require` and `assert` are both packages provided by https://github.com/stretchr/testify, the library we're using already within the codebase. 

The `require` package provides same global functions as the `assert` package, but instead of returning a boolean result they terminate the current test.

That means we stop at the first failure, instead of showing holistically all the assertions that pass / fail when this makes sense.